### PR TITLE
When source directories aren't supplied, infer

### LINKED ISSUE #9: Infer src paths
Instead of needing `-r src`, we should automatically use the src paths of the apps defined in `rebar.config`

### PR DESCRIPTION
**Pull request by maco**
Priorities:
- command line arguments
- typer config in rebar.config
- auto-detected app dirs
- src_dirs, extra_src_dirs, and sub_dirs in rebar.config
- try app, src, and lib

Fixes #9

### CODE REVIEW

**Review by elbrujohalcon — CHANGES_REQUESTED** [review]
Does this work for umbrella apps?
I would've expected that we need to use something along the lines of [dir_for_app/1](https://github.com/AdRoll/rebar3_format/blob/f3e0ffe4dca7368da396dd543fecf11207c8fa9d/src/rebar3_format_prv.erl#L60-L66) in `rebar3_format`.

**Review by elbrujohalcon — DISMISSED** [review]
I merged `main` here and fixed the tests, but now you should add tests with app dirs.
Also, I run it on my test app and…

```bash
===> typer: cannot access /src: No such file or directory
```

<details><summary>DEBUG=1 output</summary>

```bash
$ DEBUG=1 rebar3 typer --plt _build/default/rebar3_23.3.2_plt --show_exported
===> Load global config file /Users/fernandobenavides/.config/rebar3/rebar.config
===> 23.3.2 satisfies the requirement for minimum OTP version 23
===> Evaluating config script "/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_build/default/plugins/hex_core/rebar.config.script"
===> 23.3.2 satisfies the requirement for minimum OTP version 19.3
===> Compile (apps)
===> 23.3.2 satisfies the requirement for minimum OTP version 21
===> Compile (apps)
===> Running hooks for compile in app rebar3_hank (/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank) with configuration:
===> 	{pre_hooks, []}.
===> Running hooks for erlc_compile in app rebar3_hank (/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank) with configuration:
===> 	{pre_hooks, []}.
===> Analyzing applications...
===> Compiling rebar3_hank
===> compile options: {erl_opts, [warn_unused_vars,warn_export_all,
                                         warn_shadow_vars,warn_unused_import,
                                         warn_unused_function,warn_bif_clash,
                                         warn_unused_record,
                                         warn_deprecated_function,
                                         warn_obsolete_guard,
                                         strict_validation,warn_export_vars,
                                         warn_exported_vars,debug_info]}.
===> files to analyze ["/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank/src/rules/unnecessary_function_arguments.erl",
                              "/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank/src/rules/unused_callbacks.erl",
                              "/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank/src/rules/unused_record_fields.erl",
                              "/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank/src/rules/unused_macros.erl",
                              "/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank/src/rules/single_use_hrl_attrs.erl",
                              "/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank/src/rules/unused_configuration_options.erl",
                              "/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank/src/rules/unused_hrls.erl",
                              "/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank/src/rules/single_use_hrls.erl",
                              "/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank/src/rebar3_hank.erl",
                              "/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank/src/hank_context.erl",
                              "/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank/src/rebar3_hank_prv.erl",
                              "/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank/src/hank_rule.erl",
                              "/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank/src/hank.erl",
                              "/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank/src/hank_utils.erl"]
===> Running hooks for erlc_compile in app rebar3_hank (/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank) with configuration:
===> 	{post_hooks, []}.
===> Running hooks for app_compile in app rebar3_hank (/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank) with configuration:
===> 	{pre_hooks, []}.
===> Running hooks for app_compile in app rebar3_hank (/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank) with configuration:
===> 	{post_hooks, []}.
===> Running hooks for compile in app rebar3_hank (/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_hank) with configuration:
===> 	{post_hooks, []}.
===> 23.3.2 satisfies the requirement for minimum OTP version 23
===> Compile (apps)
===> Running hooks for compile in app rebar3_typer (/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_typer) with configuration:
===> 	{pre_hooks, []}.
===> Running hooks for erlc_compile in app rebar3_typer (/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_typer) with configuration:
===> 	{pre_hooks, []}.
===> Analyzing applications...
===> Compiling rebar3_typer
===> compile options: {erl_opts, [debug_info,warn_export_vars,
                                         warn_missing_specs,
                                         warn_unused_import,
                                         warnings_as_errors,
                                         {i,"src"}]}.
===> files to analyze ["/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_typer/src/rebar3_typer.erl",
                              "/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_typer/src/rebar3_mini_typer.erl",
                              "/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_typer/src/rebar3_typer_prv.erl"]
===> Running hooks for erlc_compile in app rebar3_typer (/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_typer) with configuration:
===> 	{post_hooks, []}.
===> Running hooks for app_compile in app rebar3_typer (/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_typer) with configuration:
===> 	{pre_hooks, []}.
===> Running hooks for app_compile in app rebar3_typer (/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_typer) with configuration:
===> 	{post_hooks, []}.
===> Running hooks for compile in app rebar3_typer (/Users/fernandobenavides/Projects/elbrujohalcon/test_hank/_checkouts/rebar3_typer) with configuration:
===> 	{post_hooks, []}.
===> Expanded command sequence to be run: [app_discovery,typer]
===> Running provider: app_discovery
===> Found top-level apps: [test_hank]
	using config: [{src_dirs,["src"]},{lib_dirs,["apps/*","lib/*","."]}]
===> Running provider: typer
===> Looking for types to add...
===> Opts: #{edoc => false,
                    files_r => ["/src"],
                    io =>
                        #{abort => fun rebar_api:abort/2,
                          debug => fun rebar_api:debug/2,
                          info => fun rebar_api:info/2,
                          warn => fun rebar_api:warn/2},
                    mode => show_exported,no_spec => false,
                    plt => "_build/default/rebar3_23.3.2_plt",
                    show_succ => false}
Args: {args,[],["/src"],[]}
Analysis: {analysis,show_exported,[],[],
              {codeserver,0,#Ref<0.402083263.2257190914.158999>,
                  #Ref<0.402083263.2257190914.159003>,
                  #Ref<0.402083263.2257190914.159002>,
                  #Ref<0.402083263.2257190914.159000>,
                  #Ref<0.402083263.2257190914.159001>,
                  #Ref<0.402083263.2257190914.159005>,
                  #Ref<0.402083263.2257190914.159004>,
                  #Ref<0.402083263.2257190914.159006>,
                  #Ref<0.402083263.2257190914.159007>,
                  #Ref<0.402083263.2257190914.159008>,
                  #Ref<0.402083263.2257190914.159009>},
              {callgraph,
                  {digraph,#Ref<0.402083263.2257190914.159020>,
                      #Ref<0.402083263.2257190914.159021>,
                      #Ref<0.402083263.2257190914.159022>,true},
                  undefined,#Ref<0.402083263.2257190914.159013>,
                  #Ref<0.402083263.2257190914.159017>,
                  #Ref<0.402083263.2257190914.159014>,
                  #Ref<0.402083263.2257190914.159015>,
                  #Ref<0.402083263.2257190914.159016>,
                  #Ref<0.402083263.2257190914.159018>,
                  #Ref<0.402083263.2257190914.159019>,false,<0.194.0>},
              [],"_build/default/rebar3_23.3.2_plt",false,false,false,[],
              {dict,0,16,16,8,80,48,
                  {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                  {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
              {dict,0,16,16,8,80,48,
                  {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                  {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
              {dict,0,16,16,8,80,48,
                  {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                  {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
              {dict,0,16,16,8,80,48,
                  {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                  {{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},
              {plt,#Ref<0.402083263.2257190914.159029>,
                  #Ref<0.402083263.2257190914.159031>,
                  #Ref<0.402083263.2257190914.159030>,
                  #Ref<0.402083263.2257190914.159032>,
                  #Ref<0.402083263.2257190914.159033>},
              #{abort => fun rebar_api:abort/2,debug => fun rebar_api:debug/2,
                info => fun rebar_api:info/2,warn => fun rebar_api:warn/2}}
===> Extracting trusted typer info...
===> typer: cannot access /src: No such file or directory
```
</details>

File: src/rebar3_typer_prv.erl [hunk 1]
```diff
@@ -148,6 +157,45 @@ get_plt(State) ->
...
+    end.
+
+infer_src_dirs(State) ->
+    SrcDirs = rebar_state:get(State, src_dirs, []),
+    Extra = rebar_state:get(State, extra_src_dirs, []),
+    SubDirs = rebar_state:get(State, sub_dirs, []),
+    FromState = SrcDirs ++ Extra ++ SubDirs,
+    case FromState of
+        [] ->
+            filelib:wildcard("{src,lib,apps}"); % last ditch
```
Comment (elbrujohalcon) [inline src/rebar3_typer_prv.erl:179]:
```suggestion
            filelib:wildcard("{src,lib/*/src,apps/*/src}"); % last ditch
```

File: test/dummy/rebar.config [hunk 2]
```diff
@@ -0,0 +1,4 @@
+{erl_opts, [debug_info]}.
```
Comment (elbrujohalcon) [inline test/dummy/rebar.config:1]: Let's move this file to `test/files/dummy/…`.

File: test/dummy/src/dummy.erl [hunk 3]
```diff
@@ -0,0 +1,6 @@
+-module(dummy).
```
Comment (elbrujohalcon) [inline test/dummy/src/dummy.erl:1]: This file as well.

**Review by elbrujohalcon — CHANGES_REQUESTED** [review]

**Review by elbrujohalcon — APPROVED** [review]

**Review by pablocostass — APPROVED** [review]

**Pull request merged**